### ISSUE TITLE
Bootstrapping

### ISSUE THREAD

**Issue by stefanocoretta** [stefanocoretta/speakr-demo#1]
Hi @cderv! You can check/clone the `bootstrap-quarto` [branch](https://github.com/stefanocoretta/speakr-demo/tree/bootstrap-quarto) of this repo for a bootstrapping use case example.

The file to be bootstrapped is `speaker-demo.qmd`. I set `eval: false` in the code blocks that run the Praat script after it's (supposed to be) tangled because it relies on external software (Praat) and the code blocks that plot the data that is generated by that script.

Let me know if I can provide you with more info.

**Comment by cderv** [comment]
Honestly, it is a bit too much for me to understand the project, what you are trying to do precisely, and how you do it. 

Looking at the project, it reminds me of knitr capabilities for this
* https://yihui.org/rlp/
* https://github.com/yihui/rlp/blob/master/vignettes/LP-demo1.Rmd

From afar, it just seems not right to require a chunk with `quarto_render()` - this is R specific and Quarto tools aims to be language agnostic. If you are to provide a Lua filter for that, I guess you don't wan't it to be tied to R. 

Some new stuff that could help maybe: 
- https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases 
  You can control more precisely where filters are run as part of the process. This allow to do some AST processing at various step of the rendering. 

Sorry I cannot help further on this 😞 - time is rare, unfortunately!

**Comment by stefanocoretta** [comment]
Thanks! Yes, I am trying to do something like the Literate Programming Demo of knitr. :) Actually that is very helpful! In fact, the reason for why I was resorting to running Quarto was that the script was run after Quarto was (meaning that quarto markdown is converted to pandoc markdown then the filter is run).

Of course the filter does not work on the qmd file (I tried), but I will have a look at playing with the processing phase.

If I run the Lua filter on a pandoc md file that works fine, so I wonder if there is a way of doing qmd > pandoc md and then I can just run the filter on the pandoc md file for bootstrapping. No problem if you don't have time to answer and thanks again for having looked at this!

**Comment by cderv** [comment]
> If I run the Lua filter on a pandoc md file that works fine,

I am not sure there should be a difference but not sure what you mean by Pandoc md file. Filters apply on the AST so there is no md or not. Just parsed content.  https://quarto.org/docs/extensions/filters.html#overview

Quarto will do a lot of processing on these AST nodes and the new feature allow your process to apply either before, during or after. At the end, Pandoc will write the output based on the AST provided using a writer. 

So not sure what Pandoc md means. But if your filter works on a simple .md document you render. Then it could just be a matter of adaptation for some specificities of Quarto AST processing. Or maybe just apply your filter before most of Quarto process if that can work for this. IDK. 

Hope it helps

**Comment by stefanocoretta** [comment]
Yes, it helps! I think knitr might be part of the story: if I run the lua filter on the qmd file with pandoc from the terminal, the filter does not work. But if I run it on the md file generated by knitr (i.e. knitr is run on the qmd file and an knitr.md file is generated or something like that right?) then the filter runs fine.

I assume there is a way of using knitr to just get the knitr.md file without evaluating code and then just run the filter on that using pandoc rather than quarto.

**Comment by cderv** [comment]
> if I run the lua filter on the qmd file with pandoc from the terminal, the filter does not work.

Oh now I see. - yeah .qmd file is not directly parsable by Pandoc if you use some specific Quarto feature. Filters needs to run as part of the rendering process. Quarto as a special reader for qmd in Pandoc. 

> But if I run it on the md file generated by knitr (i.e. knitr is run on the qmd file and an knitr.md file is generated or something like that right?) then the filter runs fine.

This file is the same as the one the Quarto Lua processing applies. So running your filters in the pipeline of Quarto should work, doesn't it ? 

> I assume there is a way of using knitr to just get the knitr.md 

It is not as easy because you won't get all the exact same behavior as when Quarto runs `rmarkdown::render()` / `knitr::knit()` for your in the process. 

You really should try to understand Lua processing, AST representation, understand when your filter should apply. Possibly you want to use your filter as first step in https://quarto.org/docs/prerelease/1.4/lua_changes.html#more-precise-targeting-of-ast-processing-phases

Otherwise, if the aim is to start from a .qmd to produce a .md you get try to play with custom format to provide a `format: custom` that would do what you need and apply you filter. Then you can `quarto render` that new file in another step. 🤷
